### PR TITLE
Central and Peripheral PAST samples

### LINKED ISSUE #32669: [Bluetooth] sample code for Periodic Advertising Sync Transfer
**Is your feature request related to a problem? Please describe.**
No, this is NOT related to a problem. 

**Describe the solution you'd like**
Periodic Adv Sync Transfer(PAST) is an important feature especially LE Audio, current Zephyr Bluetooth stack supports PAST(it may need the support from Bluetooth controller), there could be a sample code to illustrate this feature. 
There are several modes of PAST, we can select the typical one, three devices involved: PA, PA sender and PA receiver to do it. 

**Describe alternatives you've considered**
No. 

**Additional context**
None.

### PR DESCRIPTION
**Pull request by Thalley**
Adds two samples that work together (and together with the periodic advertising sample) to show how to use the periodic advertising sync transfer feature from Bluetooth 5.1. 

fixes https://github.com/zephyrproject-rtos/zephyr/issues/32669

### CODE REVIEW

**Review by joerchan — APPROVED** [review]
@Thalley Sample looks ok, but have you considered what happens when the central disconnects in the middle of the main loop. Wouldn't it get stuck between the k_sem_waits?

**Comment by Thalley** [comment]
> @Thalley Sample looks ok, but have you considered what happens when the central disconnects in the middle of the main loop. Wouldn't it get stuck between the k_sem_waits?

That's a good question. If it disconnects after `err = k_sem_take(&sem_conn, K_FOREVER);` then it would still wait for the PA sync and then `bt_le_per_adv_sync_transfer` would fail  as the connection pointer would be invalid, it would return, if it doesn't create a new connection before that (the `disconnected` callback starts the scan again which will attempt to connect again). It is actually worse if we lose the PA sync, as that isn't handled very well. 

It isn't trivial to ensure both features (conn and pa_sync) are in the right state all the time in a single thread. I had considered creating a separate thread handling the PA sync, such that we could have 2 while loops running, each ensuring that their respective feature is handled. 

Do you have any ideas of handling this?

**Comment by joerchan** [comment]
@Thalley Maybe you could use k_poll here instead?

**Comment by Thalley** [comment]
> @Thalley Maybe you could use k_poll here instead?

As far as I can tell, there are no places where `k_sem_take` poses an issue. The potential issue I can see are PA or conn being in a "bad" state when the PAST is performed, which causes it to return and basically stop the sample.

**Comment by joerchan** [comment]
> > @Thalley Maybe you could use k_poll here instead?
> 
> As far as I can tell, there are no places where `k_sem_take` poses an issue. The potential issue I can see are PA or conn being in a "bad" state when the PAST is performed, which causes it to return and basically stop the sample.

Fair enough, and you could always just restart the sample if something went wrong.

**Comment by Thalley** [comment]
> > > @Thalley Maybe you could use k_poll here instead?
> > 
> > 
> > As far as I can tell, there are no places where `k_sem_take` poses an issue. The potential issue I can see are PA or conn being in a "bad" state when the PAST is performed, which causes it to return and basically stop the sample.
> 
> Fair enough, and you could always just restart the sample if something went wrong.

Indeed, that was my initial thought as well. Not sure how much time we should spend on making error resilient samples.

File: samples/bluetooth/central_past/src/main.c [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+	char le_addr[BT_ADDR_LE_STR_LEN];
+
+	printk("Starting Central Periodic Advertising Synchronization Transfer (PAST) Demo\n");
+
+	/* Initialize the Bluetooth Subsystem */
+	err = bt_enable(NULL);
+
+	printk("Connection callbacks register...");
+	bt_conn_cb_register(&conn_callbacks);
+	printk("Success\n");
```
Comment (asbjornsabo) [inline samples/bluetooth/central_past/src/main.c:245]: Nitpicking: Can this ever fail?  I assume it is just to be consistent with the pattern below.

File: samples/bluetooth/peripheral_past/src/main.c [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <devicetree.h>
+#include <drivers/gpio.h>
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/conn.h>
+
+static struct bt_conn *default_conn;
```
Comment (asbjornsabo) [inline samples/bluetooth/peripheral_past/src/main.c:13]: Never gets set

**Review by asbjornsabo — CHANGES_REQUESTED** [review]
It seems that the default connection pointer never gets set in the peripheral.

File: samples/bluetooth/central_past/src/main.c (hunk 1 above)
Comment (Thalley) [reply samples/bluetooth/central_past/src/main.c:245]: It cannot. I've remove the "Success" prints

**Review by asbjornsabo — APPROVED** [review]
My comments have been addressed. LGTM

**Pull request merged**